android/src/main/java/com/swmansion/gesturehandler/core/RotationGestureDetector.kt

```diff
@@ -85,50 +120,109 @@ class RotationGestureDetector(private val gestureListener: OnRotationGestureList
+        action: ${event.action}, actionIndex: ${event.actionIndex} pointerCount: ${event.pointerCount}
+        rotation: $rotation, previousAngle: $previousAngle
+        event: $event
+      """.trimIndent()
+    }
+    localLog("onTouchEvent START | ${getLogInfo()}", true)
+
     when (event.actionMasked) {
       MotionEvent.ACTION_DOWN -> {
+        localLog("ActionDown START | ${getLogInfo()}")
         isInProgress = false
         pointerIds[0] = event.getPointerId(event.actionIndex)
         pointerIds[1] = MotionEvent.INVALID_POINTER_ID
+        localLog("ActionDown END | ${getLogInfo()}")
       }
-      MotionEvent.ACTION_POINTER_DOWN -> if (!isInProgress) {
-        pointerIds[1] = event.getPointerId(event.actionIndex)
-        isInProgress = true
-        previousTime = event.eventTime
-        previousAngle = Double.NaN
-        updateCurrent(event)
-        gestureListener?.onRotationBegin(this)
+      MotionEvent.ACTION_POINTER_DOWN -> {
+        localLog("ActionPointerDown START | ${getLogInfo()}")
+        if (!isInProgress) {
+          pointerIds[1] = event.getPointerId(event.actionIndex)
+          isInProgress = true
+          previousTime = event.eventTime
+          previousAngle = Double.NaN
+          updateCurrent(event)
+          gestureListener?.onRotationBegin(this)
+        }
+        localLog("ActionPointerDown END | ${getLogInfo()}")
       }
       MotionEvent.ACTION_MOVE -> if (isInProgress) {
-        updateCurrent(event)
-        gestureListener?.onRotation(this)
-      }
-      MotionEvent.ACTION_POINTER_UP -> {
-        val pointerId = event.getPointerId(event.actionIndex)
+        localLog("ActionMove START | ${getLogInfo()}", true)
+        val firstPointerIndex = event.findPointerIndex(pointerIds[0])
+        val secondPointerIndex = event.findPointerIndex(pointerIds[1])
 
-        // All key pointers are up
-        if (!isInProgress && pointerId == pointerIds[0]) {
-          gestureListener?.onRotationEnd(this)
+        if (
+          firstPointerIndex != MotionEvent.INVALID_POINTER_ID &&
+          secondPointerIndex != MotionEvent.INVALID_POINTER_ID
+        ) {
+          updateCurrent(event)
+          gestureListener?.onRotation(this)
         }
-
-        // One of the key pointers is up
-        if (isInProgress && pointerIds.contains(pointerId)) {
-          if (pointerId == pointerIds[0]) {
-            pointerIds[0] = pointerIds[1]
+        localLog("ActionMove END | ${getLogInfo()}", true)
+      }
+      MotionEvent.ACTION_POINTER_UP -> {
+        localLog("ActionPointerUp START | ${getLogInfo()}")
+        if (isInProgress) {
+          val pointerId = event.getPointerId(event.actionIndex)
+          if (pointerId == pointerIds[0] || pointerId == pointerIds[1]) {
+            // One of the key Pointers has been lifted up, we have to pause or finish the Gesture
+            if (secondPointerLiftFinishesGesture) finish()
+            else pause()
           }
```

Comment:
I think we should keep moving the pointer id to the first index if the gesture is in progress so we don't store a dangling id.

```kotlin
if (pointerId == pointerIds[0]) {
  pointerIds[0] = pointerIds[1]
}
pointerIds[1] = MotionEvent.INVALID_POINTER_ID
```

Reply:
If we simply toss out second Pointer on any `MotionEvent.ACTION_POINTER_UP` – Gesture will break on a non-key Pointer Lift (eg: third Finger)

I suggest tossing (replacing stored ID with `MotionEvent.INVALID_POINTER_ID`) exactly lifted Pointer

Reply:
Ah, you're right. You can also immediately shift the second pointer so that `pointerIds[0]` is always a valid pointer id while the gesture is being recognized.

```kt
if (pointerId == pointerIds[0]) {
  pointerIds[0] = pointerIds[1]
  pointerIds[1] = MotionEvent.INVALID_POINTER_ID
  tryPause()
} else if (pointerId == pointerIds[1]) {
  pointerIds[1] = MotionEvent.INVALID_POINTER_ID
  tryPause()
}
```

Reply:
Sounds good, thank you!

[Updated the PR](https://github.com/software-mansion/react-native-gesture-handler/pull/3057/commits/e5528492347e7341406d11e66a2cd8cec4cb9186)